File: AzDG.go
/**
 * Created with IntelliJ IDEA.
 * User: eddie
 * Date: 2013/3/29
 * Time: 上午 11:56
 * To change this template use File | Settings | File Templates.
 */
package eddie
import (
	"fmt"
	"crypto/md5"
	"encoding/base64"
	"time"
)

var cipher = "Private key"

var ( h  = md5.New() )

func cipherEncode(sourceText string) string {
	h.Write([]byte(cipher))
	cipherHash := fmt.Sprintf("%x", h.Sum(nil))
	h.Reset()
	inputData := []byte(sourceText)
	loopCount := len(inputData)
	outData := make([]byte,loopCount)
	for i:= 0; i < loopCount ; i++ {
		outData[i] = inputData[i] ^ cipherHash[i%32]
	}
	return fmt.Sprintf("%s", outData)
}

func Encrypt(sourceText string) string {
	h.Write([]byte(time.Now().Format("2006-01-02 15:04:05")))
	noise := fmt.Sprintf("%x", h.Sum(nil))
	h.Reset()
	inputData := []byte(sourceText)
	loopCount := len(inputData)
	outData := make([]byte,loopCount*2)
	for i, j := 0,0; i < loopCount ; i,j = i+1,j+1 {
		outData[j] = noise[i%32]
		j++
		outData[j] = inputData[i] ^ noise[i%32]
	}
	return base64.StdEncoding.EncodeToString([]byte(cipherEncode(fmt.Sprintf("%s", outData))))
}

func Decrypt(sourceText string) string {
	buf, err := base64.StdEncoding.DecodeString(sourceText)
	if err != nil {
		fmt.Println("Decode(%q) failed: %v", sourceText, err)
		return ""
	}
	inputData := []byte(cipherEncode(fmt.Sprintf("%s", buf)))
	loopCount := len(inputData)
	outData := make([]byte,loopCount)
	for i, j := 0,0; i < loopCount ; i,j = i+2,j+1 {
		outData[j] = inputData[i] ^ inputData[i+1]
	}
	return fmt.Sprintf("%s", outData)
}

File: AzDG.dart
import 'package:crypto/crypto.dart' show MD5, CryptoUtils;
import 'package:utf/utf.dart' ;

class AzDG {
  String _cipher = 'Private key';  
  AzDG({String cipher}) {   
    if (cipher != null)
    {
      _cipher = cipher;
    }
  }
  
  List<int> _cipherEncode(List<int> inputData)
  {
    var md5 = new MD5(); 
    md5.add(this._cipher.codeUnits);
    var cipherHash = CryptoUtils.bytesToHex(md5.close()).codeUnits; 
    var outData = new List<int>(inputData.length); 
    var loopCount = inputData.length;
    for (var i = 0; i < loopCount; i++) {
      outData[i] =  inputData[i] ^ cipherHash[i%32];
    }  
    return outData;
  }
  
  String Decrypt(String sourceText)
  {
    var decodeSourceText = _cipherEncode(CryptoUtils.base64StringToBytes(sourceText));   
    var size = (decodeSourceText.length ~/ 2).toInt();
    var outData = new List<int>(size);
    for (int j = 0; j < size; ++j) {     
      outData[j] = (decodeSourceText[(j*2)] ^ decodeSourceText[(j*2)+1]);
    }
    return decodeUtf8(outData);
  }
  
  String Crypt(String sourceText)
  {
    var now = new DateTime.now();
    var md5 = new MD5();   
    md5.add(now.toString().codeUnits);
    var noise = CryptoUtils.bytesToHex(md5.close()).codeUnits;
    var inputData = encodeUtf8(sourceText);
    var loopCount = inputData.length;
    var outData = new List<int>(loopCount*2);
    for (var j = 0; j < loopCount ; j++) {
        outData[(j*2)] = noise[j%32];      
        outData[(j*2)+1] = inputData[j] ^ noise[j%32];
    }
    return CryptoUtils.bytesToBase64(_cipherEncode(outData));
  }
}
